### DISCUSSION TITLE
Variadics in patterns

### DISCUSSION THREAD

**Discussion by Kixiron** [egraphs-good/egg#195] [Q&A]
I'm wanting to model uninterpreted functions in egg but I'm not sure how I'd do something akin to this:
```lisp
(fun foo a b c)
```
Which would have the variant type of
```
enum Func {
    "fun" = Function(Symbol, Vec<Id>), // Or even just a `Vec<Id>` where the first is always a `Symbol`
}
```
The problem I forsee is trying to match on that, e.g.
```
(ne (fun ?f ?x ?y ?z) (fun ?f ?x ?y ?z)) => false
```
The problem is the trailing arguments, do you have to generate patterns for a set number of cases or is there some sort of `...?args` notation?

**Comment by mwillsey** [comment]
There are no variadic patterns in egg right now. The difficulty would be that a variable could then refer to any number of eclass ids.

Could you try a cons-list representation? That should allow you to efficiently manipulate heads of lists.

**Reply by saulshanabrook** [reply]
Would you be open to a PR which added variadic support? The cons list approach does work, but is a bit more verbose and expensive.

**Reply by mwillsey** [reply]
At this moment, I'm not sure variadics carry their weight. They open up a ton of questions for patterns and e-matching (can you partially bind variadics), and I'm just not sure it's worth the complexity. We also have plans to move to a more efficient relational e-matching algorithm at some point, and that simply doesn't support variadics.

**Reply by saulshanabrook** [reply]
@mwillsey sounds good, thank you for the explanation!
